text/src/autogluon/text/automm/models/ft_transformer.py

```diff
@@ -0,0 +1,594 @@
+            if initialization == 'xavier' and (
+                m is not self.W_v or self.W_out is not None
+            ):
+                # gain is needed since W_qkv is represented with 3 separate layers (it
+                # implies different fan_out)
+                nn.init.xavier_uniform_(m.weight, gain=1 / math.sqrt(2))
+            if m.bias is not None:
+                nn.init.zeros_(m.bias)
+        if self.W_out is not None:
+            nn.init.zeros_(self.W_out.bias)
+
+    def _reshape(self, x: Tensor) -> Tensor:
+        batch_size, n_tokens, d = x.shape
+        d_head = d // self.n_heads
+        return (
+            x.reshape(batch_size, n_tokens, self.n_heads, d_head)
+            .transpose(1, 2)
+            .reshape(batch_size * self.n_heads, n_tokens, d_head)
+        )
+
+    def forward(
+        self,
+        x_q: Tensor,
+        x_kv: Tensor,
+        key_compression: Optional[nn.Linear],
+        value_compression: Optional[nn.Linear],
+    ) -> Tuple[Tensor, Dict[str, Tensor]]:
+        """Perform the forward pass.
+
+        Args:
+            x_q: query tokens
+            x_kv: key-value tokens
+            key_compression: Linformer-style compression for keys
+            value_compression: Linformer-style compression for values
+        Returns:
+            (tokens, attention_stats)
+        """
+        assert _all_or_none(
+            [key_compression, value_compression]
+        ), 'If key_compression is (not) None, then value_compression must (not) be None'
+        q, k, v = self.W_q(x_q), self.W_k(x_kv), self.W_v(x_kv)
+        for tensor in [q, k, v]:
+            assert tensor.shape[-1] % self.n_heads == 0, _INTERNAL_ERROR_MESSAGE
+        if key_compression is not None:
+            k = key_compression(k.transpose(1, 2)).transpose(1, 2)
+            v = value_compression(v.transpose(1, 2)).transpose(1, 2)  # type: ignore
+
+        batch_size = len(q)
+        d_head_key = k.shape[-1] // self.n_heads
+        d_head_value = v.shape[-1] // self.n_heads
+        n_q_tokens = q.shape[1]
+
+        q = self._reshape(q)
+        k = self._reshape(k)
+        attention_logits = q @ k.transpose(1, 2) / math.sqrt(d_head_key)
+        attention_probs = F.softmax(attention_logits, dim=-1)
+        if self.dropout is not None:
+            attention_probs = self.dropout(attention_probs)
+        x = attention_probs @ self._reshape(v)
+        x = (
+            x.reshape(batch_size, self.n_heads, n_q_tokens, d_head_value)
+            .transpose(1, 2)
+            .reshape(batch_size, n_q_tokens, self.n_heads * d_head_value)
+        )
+        if self.W_out is not None:
+            x = self.W_out(x)
+        return x, {
+            'attention_logits': attention_logits,
+            'attention_probs': attention_probs,
+        }
+
+
+class Transformer(nn.Module):
+    """Transformer with extra features.
+
+    This module is the backbone of `FTTransformer`."""
```

Comment:
We should be careful here. Is there any open source repo that has already implemented this model that we could simply import? Do we really want to be the ones maintaining this model's logic and implementation?

Reply:
For example, FT-Transformer is already implemented in https://github.com/jrzaurin/pytorch-widedeep

(@gradientsky has been working on adding these models already)

Reply:
By implementing the model in AutoMM, we are able to support more advanced ways for encoding tabular data in our multi-modal fusion network. This can help build more powerful fusion models together with TIMM and huggingface/transformers. Also, the backend for training these models got unified with PT-Lightning and it's easy to support HPO.

Reply:
Why not just override the parts we want to change vs importing the whole codebase?